script.js

```diff
@@ -1 +1,143 @@
+  }  // setItem(key,value) 특정 key에 해당 value 할당 
+
+  function loadStorage() {
+      const storedTodos = localStorage.getItem("todos"); // 기존에 localStorage에 저장되어있던 배열을 불러오기, 만약 없다면 null이 저장 됨 
+      if (storedTodos) {
+          todoList = JSON.parse(storedTodos);
+          todoList.forEach((todo) => displayTodo(todo.text, todo.completed)); // todoList 배열을 순회하며 저장된 모든 todo를 화면에 렌더링
+      }
+      updateCounts(); // 새로 고침 시 기존 todoList 배열 불러와서 총 개수 맞게 렌더링
+  }
+
+  function displayTodo(todoText, isCompleted) {
+      const li = document.createElement("li"); // 새로운 <li>요소 생성. 하나의 todo를 나타냄
+
+      const checkbox = document.createElement("input"); // 새로운 input 요소 checkbox 생성
+      checkbox.type = "checkbox"; // 이 요소의 type = checkbox
+      checkbox.classList.add("todoCheckBox"); // 스타일링을 위해 클래스 목록에 클래스 이름 추가! 즉 checkbox에 할당되는 클래스 이름이 todo-checkbox
+      checkbox.checked = isCompleted; // isCompleted가 true라면 checkbox가 체크 됨
+
+      checkbox.addEventListener("change", function () { // checkbox의 상태가 바뀔 때, 즉 checked의 속성이 변경될 때 (사용자가 체크박스를 체크하거나 해제할 때) 실행될 이벤트 핸들러
+          toggleTodoCompletion(todoText);
+          li.querySelector("span").style.textDecoration = checkbox.checked ? "line-through" : "none";
+      }); // <li> 요소의 첫 번째 <span> 요소를 찾아 스타일링. 만약 체크박스가 체크 되어있으면? 선을 긋고 해제되어 있으면 선을 없앤다!
+
+      const todoSpan = document.createElement("span"); //<li> 요소 안에 쓰여질 텍스트를 <span> 요소로 설정
+      todoSpan.classList.add("todoSpan");
+      todoSpan.textContent = todoText; // 매개변수로 전달 받은 input의 text를 변수 todoSpan의 텍스트 내용으로 저장
+      if (isCompleted) {
+          todoSpan.style.textDecoration = "line-through";
+      }
+
+      const deleteBtn = document.createElement("button"); // 새로운 <button> 요소 생성
+      deleteBtn.textContent = "삭제"; // deleteBtn의 텍스트 내용을 삭제라고 지정
+      deleteBtn.classList.add("deleteBtn");
+      deleteBtn.addEventListener("click", function () { // 버튼이 클릭 되었을 때 실행될 함수!!
+        if(confirm("정말 삭제하시겠습니까?")){
+            alert("todo가 삭제되었습니다.");
+            deleteTodo(todoText, li);
+        }else{
+            alert("Kepp going!🔥");
+        }
+      });
+
+      li.append(checkbox, todoSpan, deleteBtn); // appendChild를 append로 간소화!
+      
+      todoBox.appendChild(li); // <ul> 요소를 불러 온 todoBox에 <li>들 추가
+  }
+
+  function toggleTodoCompletion(todoText) { // 배열 todoList에서 특정 할 일 항목의 완료 상태를 토글시키는 역할
+      todoList = todoList.map((todo) => { // map 메서드는 배열의 각 항목을 변형하여 새로운 배열을 반환한다!!
+          if (todo.text === todoText) { // 순회하던 배열의 특정 원소의 text와 display 함수에서 props로 전달 받은 체크 상태가 변경된 todo의 text가 동일하다면?
+              return { ...todo, completed: !todo.completed }; // 그 원소의 completed 상태 토글
+          }
+          return todo;
+      });
+      saveStorage(); // todoList가 업데이트 되어 새롭게 localStorage에 저장
+      updateCounts(); // 체크 표시 상태에 따라 완료된 todo 개수 변경
+  }
```

Comment:
toggleTodoCompletion 함수를 보니까 map 메서드를 통해 모든 todolist를 순회하면서 todo의 상태를 바꾸고 있는 구조인 것 같아요. 물론 이 방식도 유효하지만 find() 메서드를 사용하여 좀 더 효율적으로 코드를 수정할 수 있을 것 같아요. 대략적으로 아래와 같이 바꾸면 어떨까요?

```suggestion
function toggleTodoCompletion(todoText) { 
    const todo = todoList.find((todo) => todo.text === todoText); 

    if (todo) { 
        todo.completed = !todo.completed; 
        saveStorage(); 
        updateCounts(); 
    }
}
```

map() 메서드는 배열의 모든 요소를 순회하면서 각 요소를 변형한 새로운 배열을 반환하는데, 이번 바닐라JS 투두리스트에서 완료 상태를 토글하는 작업은 굳이 배열 전체를 변형할 필요가 없을 것 같습니다. find() 메서드는 배열을 순회하다가 첫번째로 조건을 만족하는 요소를 찾으면 그 요소를 반환하고 순회를 멈추기 때문에 하나의 요소만 업데이트해야하는 토글 함수에 더 효율적인 방향이라고 생각합니다..! 가독성 면에서도 짧고 간단해서 보기가 수월할 것 같구요..!

map()은 일반적으로 불변성을 유지하기 위해 새로운 배열을 반환하는 상황에서 많이 사용된다고 알고 있는데, (제 생각으로는) 이 부분의 코드에서는 불변성 유지가 필수적이지 않은 것 같아요. 만약 리액트와 같은 상태 관리 시스템에서 개발을 한다면 어디서든 불변성을 유지하는 것이 매우 중요하지만, (리액트는 상태변화가 있을 때 렌더링을 트리거하기 때문에 불변성을 지켜야 변경된 상태만 효율적으로 감지할 수 있기 때문) 로컬 스토리지에 업데이트하는 단순한 데이터 처리에서는 불변성을 유지하지 않아도 성능에 문제가 생길 확률이 적습니다..!

[불변성 유지 여부에 따른 코드 전개](https://ggcoding.tistory.com/118)